### PR TITLE
Add support for streaming on recent Darwin platforms

### PR DESCRIPTION
**Pull request by simonjbeaumont**
### Motivation

Recently, the `ClientTransport` protocol was updated to be in terms of `HTTPBody`, which is an `AsyncSequence<ArraySlice<UInt8>>`. Until now, the URLSession transport has buffered the request and response bodies in memory. This PR seeks to bring streaming support to recent Darwin platforms.

### Modifications

On new enough Darwin platforms[^1], use URLSession delegate to perform bidirectional streaming with backpressure. Request body backpressure is provided by the use of `StreamDelegate` with a bound stream pair. Response body backpressure is provided by the use of `AsyncBackpressuredStream` which allows for a high and low watermark to suspend and resume the URLSession task.

[^1]: macOS 12, iOS 15, tvOS 15, watchOS 8

In more detail:

- Vendor internal `AsyncBackpressuredStream` implementation from SE-0406.
- Add `HTTPBodyOutputStreamBridge` which provides the `OutputStreamDelegate` required to bridge the `AsyncSequence`-based `HTTPBody` to an `OutputStream`.
- Add `BidirectionalStreamingURLSessionDelegate` which makes use of `HTTPBodyOutputStreamBridge` for streaming request bodies and also uses the delegate callbacks to bridge the response body to `HTTPBody`.
- Add `URLSession.bidirectionalStreamingRequest(...) async throws`, which provides a high-level URLSession-like API for setting up the appropriate URLSession task and delegate.
- Add `URLSession.bufferedRequest(...) async throws` to provide a similar interfaces on platforms that don't support the streaming APIs on which we depend.
- Add internal `enum URLSessionTransport.Configuration.Implementation` to control whether to use buffered or streaming implementation.
- Detect appropriate implementation depending on the platform.

### Result

- On new enough Darwin platforms[^1], streaming will be used.

### Test Plan

- Add a set of tests that run with both buffered and streaming implementation on platforms that support streaming.
- Add a set of tests that only run on platforms that support streaming, to test request flows only possible with streaming.

However, it's worth noting that our CI only runs on Linux, so we won't be testing the majority of this new feature in CI.

### CODE REVIEW

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+            switch try responseBodyStreamSource.write(contentsOf: [ArraySlice(data)]) {
+            case .produceMore: break
+            case .enqueueCallback(let writeToken):
+                if self.hasSuspendedURLSessionTask {
+                    debug("Task delegate: already suspended task, not enqueing another writer callback")
+                    break
+                }
+                debug("Task delegate: response stream backpressure, suspending URLSession task and enqueing callback")
+                dataTask.suspend()
+                self.hasSuspendedURLSessionTask = true
```
Comment (czechboy0) [inline Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift:121]: What's the threading story here? Do we not need locks because the delegate calls are guaranteed to be called on a serial queue?

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+                responseBodyStreamSource.enqueueCallback(writeToken: writeToken) { result in
+                    switch result {
+                    case .success:
+                        debug("Task delegate: response stream callback, resuming URLSession task")
+                        dataTask.resume()
+                        self.hasSuspendedURLSessionTask = false
+                    case .failure(let error):
+                        debug("Task delegate: response stream callback, cancelling URLSession task, error: \(error)")
```
Comment (czechboy0) [inline Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift:129]: Will we surface this error to the adopter?

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+import OpenAPIRuntime
+import HTTPTypes
+#if canImport(Darwin)
+import Foundation
+#if canImport(FoundationNetworking)
+@preconcurrency import struct FoundationNetworking.URLRequest
+@preconcurrency import class FoundationNetworking.URLSession
+@preconcurrency import class FoundationNetworking.URLResponse
+@preconcurrency import class FoundationNetworking.HTTPURLResponse
+#endif
```
Comment (czechboy0) [inline Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift:23]: Do we need this if we're inside the `#if canImport(Darwin)` here?

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/HTTPBodyOutputStreamBridge.swift [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+import OpenAPIRuntime
+import HTTPTypes
+#if canImport(Darwin)
+import Foundation
+#if canImport(FoundationNetworking)
+@preconcurrency import struct FoundationNetworking.URLRequest
+@preconcurrency import class FoundationNetworking.URLSession
+@preconcurrency import class FoundationNetworking.URLResponse
+@preconcurrency import class FoundationNetworking.HTTPURLResponse
+#endif
```
Comment (czechboy0) [inline Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/HTTPBodyOutputStreamBridge.swift:23]: Same as in the previous file, will these Linux imports be used if we're `#if` on Darwin?

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/HTTPBodyOutputStreamBridge.swift [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+            performAction(self.state.errorOccurred(outputStream.streamError!))
+        case let written where written > 0:
+            debug("Output stream delegate wrote \(written) bytes to stream.")
+            performAction(
+                self.state.wroteBytes(
+                    numBytesWritten: written,
+                    streamStillHasSpaceAvailable: outputStream.hasSpaceAvailable
+                )
+            )
+        default: preconditionFailure()
```
Comment (czechboy0) [inline Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/HTTPBodyOutputStreamBridge.swift:106]: Can you add a string to the crash (here and in the rest of the file)?

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/HTTPBodyOutputStreamBridge.swift [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+        }
+    }
+
+    func stream(_ stream: Stream, handle event: Stream.Event) {
+        dispatchPrecondition(condition: .onQueue(Self.streamQueue))
+        debug("Output stream delegate received event: \(event).")
+        switch event {
+        case .openCompleted:
+            guard case .initial = state else {
+                debug("Output stream delegate ignroing duplicate openCompleted event.")
```
Comment (czechboy0) [inline Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/HTTPBodyOutputStreamBridge.swift:116]:
```suggestion
                debug("Output stream delegate ignoring duplicate openCompleted event.")
```

File: Sources/OpenAPIURLSession/URLSessionTransport.swift [hunk 7]
```diff
@@ -73,8 +73,21 @@ public struct URLSessionTransport: ClientTransport {
...
         /// - Parameter session: The URLSession used for performing HTTP operations.
-        ///     If none is provided, the system uses the shared URLSession.
-        public init(session: URLSession = .shared) { self.session = session }
+        ///   If none is provided, the system uses the shared URLSession.
+        public init(session: URLSession = .shared) { self.init(session: session, implementation: .platformDefault) }
+
+        enum Implementation {
+            case buffered
+            case streaming(requestBodyStreamBufferSize: Int, responseBodyStreamWatermarks: (low: Int, high: Int))
+        }
```
Comment (czechboy0) [inline Sources/OpenAPIURLSession/URLSessionTransport.swift:82]: 👍

File: Sources/OpenAPIURLSession/URLSessionTransport.swift [hunk 8]
```diff
@@ -73,8 +73,21 @@ public struct URLSessionTransport: ClientTransport {
...
+
+        var implemenation: Implementation
+
+        init(session: URLSession = .shared, implementation: Implementation = .platformDefault) {
```
Comment (czechboy0) [inline Sources/OpenAPIURLSession/URLSessionTransport.swift:86]:
```suggestion
        init(session: URLSession = .shared, implementation: Implementation) {
```

This way `.init()` won't be ambiguous.

File: Sources/OpenAPIURLSession/URLSessionTransport.swift [hunk 9]
```diff
@@ -205,6 +218,69 @@ extension URLSessionTransportError: CustomStringConvertible {
...
+        }
+        return (try HTTPResponse(response), maybeResponseBody)
+    }
+}
+
+extension URLSessionTransport.Configuration.Implementation {
+    static var platformSupportsStreaming: Bool {
+        #if canImport(Darwin)
+        guard #available(macOS 12, iOS 15, tvOS 15, watchOS 8, *) else { return false }
+        _ = URLSession.bidirectionalStreamingRequest
```
Comment (czechboy0) [inline Sources/OpenAPIURLSession/URLSessionTransport.swift:272]: Why is this here?

**Review by czechboy0 — COMMENTED** [review]
Looks great on my first pass, left a few questions. I'll take one more look once the NIO changes landed and this is finalized.

File: Sources/OpenAPIURLSession/URLSessionTransport.swift (hunk 9 above)
Comment (simonjbeaumont) [reply Sources/OpenAPIURLSession/URLSessionTransport.swift:276]: It's here to just to check that the availability guard we used above is the right one. Given this is just a computed boolean property, there's nothing that prevented me getting the version numbers in the guard wrong.

File: Sources/OpenAPIURLSession/URLSessionTransport.swift (hunk 8 above)
Comment (simonjbeaumont) [reply Sources/OpenAPIURLSession/URLSessionTransport.swift:86]: I'm not seeing any ambiguity issues when calling `URLSessionTransport.Configuration.init()` in the tests, even when `@testable import OpenAPIURLSession`. Note that this initializer is `internal`, if that makes any difference to the ambiguity story.

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/HTTPBodyOutputStreamBridge.swift (hunk 5 above)
Comment (simonjbeaumont) [reply Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/HTTPBodyOutputStreamBridge.swift:106]: Added message to all the `precondition` and `preconditionFailure` calls throughout (except a few in tests).

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift (hunk 2 above)
Comment (simonjbeaumont) [reply Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift:129]: It will surface at the HTTP request failed and the `ClientError.underlyingError` will have a `URLError` with operation cancelled IIUC.

We have the following tests:
- `testHTTPBodyOutputStreamBridgeStreamClosedImmediately`
- `testHTTPBodyOutputStreamBridgeStreamClosedEarly`

But we don't have a test at the URLSession level here. 

I can write a test that's independent of all this logic to show what surface in the case the URLSessionTask gets cancelled if you like? Such a test would provide a session delegate that always cancels the task and we can see what error comes out of the `URLSessionTransport`.

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift (hunk 1 above)
Comment (simonjbeaumont) [reply Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift:121]: My understanding was that delegate callbacks were called on the delegate queue.

However, I just put in some `dispatchPrecondition(condition: .onQueue(session.delegateQueue.underlyingQueue!))` and it appears not.

This is an important thing we need to get to the bottom of before merging this PR.

However, we should also consider what happens when this callback fires.

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+        )
+    }
+
+    func urlSession(_ session: URLSession, needNewBodyStreamForTask task: URLSessionTask) async -> InputStream? {
+        debug("Task delegate: needNewBodyStreamForTask")
+        // If the HTTP body cannot be iterated multiple times then bad luck; the only thing
+        // we can do is cancel the task and return nil.
+        if hasAlreadyIteratedRequestBody {
+            guard requestBody!.iterationBehavior == .multiple else {
+                debug("Task delegate: Cannot rewind request body, canceling task")
```
[the remark this answers is not in the document]
Comment (simonjbeaumont) [reply Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift:93]: Yes, as discussed in the other thread.

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+            )
+        )
+    }
+
+    func urlSession(_ session: URLSession, needNewBodyStreamForTask task: URLSessionTask) async -> InputStream? {
+        debug("Task delegate: needNewBodyStreamForTask")
+        // If the HTTP body cannot be iterated multiple times then bad luck; the only thing
+        // we can do is cancel the task and return nil.
+        if hasAlreadyIteratedRequestBody {
+            guard requestBody!.iterationBehavior == .multiple else {
```
[the remark this answers is not in the document]
Comment (simonjbeaumont) [reply Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift:92]: Yes: the `urlSession(_:needNewBodyStreamForTask:)` callback is only called when using `URLSession.uploadTask(withStreamedRequest:)`, which, in turn, is only used when `requestBody` is not-nil.

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift (hunk 2 above)
Comment (czechboy0) [reply Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift:129]: Doing a one-off test is also fine, I'm just curious whether the adopter gets enough information to start debugging the failure; that we're not dropping any diagnostics here.

File: Sources/OpenAPIURLSession/URLSessionTransport.swift (hunk 8 above)
Comment (czechboy0) [reply Sources/OpenAPIURLSession/URLSessionTransport.swift:86]: Not sure, but if it's compiling fine, then we can keep it. I'll have to catch up on the precedence rules in Swift, because I'm not sure which one will actually be called in practice.

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift (hunk 2 above)
Comment (simonjbeaumont) [reply Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift]: Added a test for this. `send` fails with `URLError` with `errorCode == .cancelled`. This should end up further up the stack as a `ClientError`.

Do we think this is enough, or would we like an explicit (name TDB) `URLSessionTransportError.cannotReplayResponseBody`?

Also, this interplays with the discussion on whether we do something smarter when given a request body with single iteration behavior.

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift (hunk 3 above)
Comment (simonjbeaumont) [reply Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift]: Removed.

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift (hunk 2 above)
Comment (czechboy0) [reply Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift]: What does the `error` hold that is being provided by `case .failure(let error):`? Is that relevant to the user?

Comment (simonjbeaumont) [reply Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift]: This is coming from the `AsyncBackpressuredStream` and IIUC (@FranzBusch can confirm) it means the consumer went away, i.e. the iterator went out of scope. I _think_ therefore, it means they didn't finish consuming the response body for some reason. Presumably, in this case, they're not interested in the error by that time? 

It gets returned by the stream if the `action` after enqueuing the producer is `resumeProducerWithCancellationError`, and it always returns `CancellationError()`:

```swift
            case .resumeProducerWithCancellationError(let onProduceMore): onProduceMore(.failure(CancellationError()))
````

Comment (czechboy0) [reply Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift]: Cool, yeah then we can just always represent this by a cancelation error.

**Review by czechboy0 — APPROVED** [review]
Ok lgtm

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/HTTPBodyOutputStreamBridge.swift [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+final class HTTPBodyOutputStreamBridge: NSObject, StreamDelegate {
+    static let streamQueue = DispatchQueue(label: "HTTPBodyStreamDelegate", autoreleaseFrequency: .workItem)
+
+    let httpBody: HTTPBody
+    let outputStream: OutputStream
+    private(set) var state: State {
+        didSet { debug("Output stream delegate state transition: \(oldValue) -> \(state)") }
+    }
+
+    init(_ outputStream: OutputStream, _ httpBody: HTTPBody, openOutputStream: Bool = true) {
```
Comment (guoye-zhang) [inline Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/HTTPBodyOutputStreamBridge.swift:28]: Is `openOutputStream` still useful? I think it's always true now

File: Sources/OpenAPIURLSession/URLSessionTransport.swift [hunk 13]
```diff
@@ -73,8 +73,23 @@ public struct URLSessionTransport: ClientTransport {
...
         /// Creates a new configuration with the provided session.
         /// - Parameter session: The URLSession used for performing HTTP operations.
-        ///     If none is provided, the system uses the shared URLSession.
-        public init(session: URLSession = .shared) { self.session = session }
+        ///   If none is provided, the system uses the shared URLSession.
+        public init(session: URLSession = .shared) { self.init(session: session, implementation: .platformDefault) }
+
+        enum Implementation {
+            case buffering
+            case streaming(requestBodyStreamBufferSize: Int, responseBodyStreamWatermarks: (low: Int, high: Int))
```
Comment (guoye-zhang) [inline Sources/OpenAPIURLSession/URLSessionTransport.swift:81]: My preference is to not expose `requestBodyStreamBufferSize`. It is an extremely confusing parameter that does not reflect buffering done in CFNetwork and Network framework. We could have gotten away with not using bounded stream pair if NSStream subclassing were easier in Swift, and it will likely be unnecessary in the future.

We'd like to retain the ability to fine tune these values on our platforms without needing developers to change their code, e.g. reducing the buffer size on low memory devices like Apple Watch, or even adjusting it dynamically based on memory pressure / app state. Most developers are not equipped to decide the best buffer size to use.

Would they be better as environment variables for advanced use cases that we don't need to commit as APIs?

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+    ///
+    /// Although all the delegate callbacks are performed on the session's `delegateQueue`, there is no guarantee that
+    /// this is a _serial_ queue.
+    ///
+    /// Regardless of the type of delegate queue, URLSession will attempt to order the callbacks for each task in a
+    /// sensible way, but it cannot be guaranteed, specifically when the URLSession task is cancelled.
+    ///
+    /// Therefore, even though the `suspend()`, `resume()`, and `cancel()` URLSession methods are thread-safe, we need
+    /// to protect any mutable state within the delegate itself.
+    let callbackLock = NSLock()
```
Comment (guoye-zhang) [inline Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift:77]: Do we prefer NSLock or NIOLock? I think both are used in some files

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+
+    func urlSession(_ session: URLSession, needNewBodyStreamForTask task: URLSessionTask) async -> InputStream? {
+        callbackLock.withLock {
+            debug("Task delegate: needNewBodyStreamForTask")
+            // If the HTTP body cannot be iterated multiple times then bad luck; the only thing
+            // we can do is cancel the task and return nil.
+            if hasAlreadyIteratedRequestBody {
+                guard requestBody!.iterationBehavior == .multiple else {
+                    debug("Task delegate: Cannot rewind request body, cancelling task")
+                    task.cancel()
```
Comment (guoye-zhang) [inline Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift:105]: Returning nil is sufficient, no need to do both

**Review by guoye-zhang — APPROVED** [review]

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift (hunk 14 above)
Comment (czechboy0) [reply Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift:77]: This library doesn't depend on NIO.

File: Sources/OpenAPIURLSession/URLSessionTransport.swift (hunk 13 above)
Comment (simonjbeaumont) [reply Sources/OpenAPIURLSession/URLSessionTransport.swift:81]: > My preference is to not expose `requestBodyStreamBufferSize`. It is an extremely confusing parameter that does not reflect buffering done in CFNetwork and Network framework. 
> ...
> We'd like to retain the ability to fine tune these values on our platforms without needing developers to change their code, e.g. reducing the buffer size on low memory devices like Apple Watch, or even adjusting it dynamically based on memory pressure / app state. Most developers are not equipped to decide the best buffer size to use.

This is _not_ exposed. This `enum Implementation` is an internal type, only used in the `internal init(session:implementation:)`. The only public initializer is `public init(session:)` which calls `internal init(session:implementation:)` with `Implementation.platformDefault`, which has some hard-coded buffer sizes and is also internal, which could allow us to tweak on a per-platform basis.

> We could have gotten away with not using bounded stream pair if NSStream subclassing were easier in Swift, and it will likely be unnecessary in the future.

Oh, that's interesting to know. I'll watch this space 👀 

> Would they be better as environment variables for advanced use cases that we don't need to commit as APIs?

Hadn't considered an environment variable based approach. I think we could incorporate that too, for advanced users.

One thing to call out here that we were _considering_ exposing was a way to _force_ buffering. Right now, when providing an `HTTPBody` with `iterationBehavior == .single`, the observed behaviour is different when faced with a HTTP redirect. Specifically, it will succeed when using buffering, but _fail_ when using streaming, because we cannot rewind the request body stream.

We have other ideas on how we could solve this—e.g. wrapping the provided `AsyncSequence` in another that buffers the first N bytes and might have a better chance of replaying on redirect—but in general we may want to expose the ability to just turn off streaming.

Curious if you have any thoughts on this?

//cc @dnadoba @czechboy0

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift (hunk 15 above)
Comment (simonjbeaumont) [reply Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift:105]: Hm, I just tried removing `task.cancel()` here and `testHTTPRedirect_singleIterationBehavior_fails` never completes.

Without the `task.cancel()`, the `needNewBodyStreamForTask` callback is called in a loop.

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift (hunk 14 above)
Comment (simonjbeaumont) [reply Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift:77]: No, but it _has_ vendored in NIOLock for the AsyncBackpressuredStream. Let me see if I can switch to it.

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/HTTPBodyOutputStreamBridge.swift (hunk 12 above)
Comment (simonjbeaumont) [reply Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/HTTPBodyOutputStreamBridge.swift:28]: I think we _used_ to _not_ open it in tests where we wanted some finer grain control, but I can see that you're right: it is always true now—will remove, thanks!

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift (hunk 14 above)
Comment (simonjbeaumont) [reply Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift]: Switched to `NIOLock` in 84dd06e.

File: Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/HTTPBodyOutputStreamBridge.swift (hunk 12 above)
Comment (simonjbeaumont) [reply Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/HTTPBodyOutputStreamBridge.swift]: Removed `openOutputStream` initializer parameter in 9bb44eb.

File: Sources/OpenAPIURLSession/URLSessionTransport.swift (hunk 13 above)
Comment (czechboy0) [reply Sources/OpenAPIURLSession/URLSessionTransport.swift:81]: Maybe let's start with the parameter being internal, and only make it public if a genuine use case comes up.

I'm not a fan of libraries checking environment variables, it makes them difficult to test and reason about.
